lib/index.js

```diff
@@ -1,8 +1,76 @@
 import fetch from 'node-fetch';
-import { promises as fs } from 'fs';
+import fs from 'fs';
 
-fetch('https://www.reddit.com/r/pics/search.json?q=programming&sort=new')
-  .then(res => res.json())
-  .then((res) => fs.writeFile('data.json', JSON.stringify(res)))
-  .then(() => console.log('Saved Successfully'))
-  .catch(err => console.error(err));
\ No newline at end of file
+// Set the fetch -> json into a helper function instead of repeating them
+
+export const searchRedditPostsByTopicPromise = (topic, fields = [], sort = "new") => {
+
+  const url = getRedditUrl(topic, sort);
+
+  fetchData(url)
+    .then(res => processData(res, fields))
+    .then(posts => saveResults('data.json', posts))
+    .catch(err => console.error('Error: ', err));
+}
+
+export const searchRedditPostsByTopicAsync = async (topic, fields = [], sort = "new") => {
+
+  const url = getRedditUrl(topic, sort);
+  
+  try { 
+    const data = await fetchData(url);
+    const posts = processData(data, fields);
+
+    saveResults('data.json', posts);
+  } catch(err) {
+    console.error(err);
+  }
+}
+
+// Pass all the data into specify fields
+   // Extract Fields
+   // Allow it to be specific to an array of objects
+   // Could add a function Process data and use specify/extract fields in it
+   // In process data do we want it allow the developer to choose what 
+
+
+export const fetchData = (url) => {
+  return fetch(url)
+    .then(res => res.json())
+    .catch(err => console.error(err))
+}
+
+export const processData = (res, fields) => {
+  const data = res?.data?.children || [];
+  
+  return extractFields(data, fields);
+}
+
+export const extractFields = (data, fields) => {
+  if (!fields.length || !fields) return data;
+
+  return data.map(obj => {
+    if (typeof obj === 'object') {
+      return fields.reduce((acc, field) => {
+        if (obj?.data[field] !== undefined) {
+          acc[field] = obj?.data[field];
+        }
+  
+        return acc;
+      }, {});
+    }
+  });
+
+}
+
+export const getRedditUrl = (topic, sort = "new") => {
+  if (!topic && typeof topic !== 'string') return new Error('Topic not a string!');
+  if (typeof sort !== 'string') return new Error('Sort is not a string!'); 
+
+  return `https://www.reddit.com/r/pics/search.json?q=${topic}&sort=${sort}`;
+}
+
+export const saveResults = (filename, data) => {
```

Comment:
I like that we factored this out. Nice work!

### What is this functions single responsibility?
It appears the single responsibility is to write a JSON file

### Question
Can we rename this function to match it's single responsibility?